# Тести для малих вибірок та пропорцій



**Завдання 1**. E-commerce компанія після редизайну сайту підозрює, що **середній час до покупки (time-to-purchase)** користувачів **збільшився**.
Історично середній час від першого заходу на сайт до покупки становив $\mu_0 = 123.7$ хвилин.

Команда аналітиків випадково обрала дані **7 покупців після редизайну**:
`128, 135, 121, 142, 126, 151, 123`

З допомогою наявних даних зʼясуйте, чи збільшився середній час покупки після редизайну.

Для цього виконайте наступні 6 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть нульову та альтернативну гіпотези і визначте тип тесту.
2. Обчисліть вибіркові статистики: $\bar x$, $s$, $n$.
3. Оберіть тип тесту та виконайте його (знайдіть критичне значення тесту, статистику тесту та р-значення) будь-яким пасуючим способом, щоб перевірити гіпотезу на рівні значущості $\alpha = 0.10$.
4. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$ використовуючи p-value.
5. Напишіть висновок: чи справді редизайн сайту подовжив час до покупки?
6. Чи зміниться ваше рішення при зміні рівня значущості на $\alpha = 0.05$.


### Маємо правосторонній тест

**Гіпотези:**  

H₀: μ = 123.7 (середній час до покупки не збільшився)  
H₁: μ > 123.7 (середній час до покупки збільшився)

In [1]:
import numpy as np
from scipy import stats

In [2]:
# Дані:
times = np.array([128, 135, 121, 142, 126, 151, 123])
mu0 = 123.7
alpha = 0.1

### Обчислимо вибіркову статистику: $\bar x$, $s$, $n$.

In [3]:
x_bar = np.mean(times)
s = np.std(times, ddof=1)
n = len(times)

### Для правостороннього тесту обчислимо: t-статистику, t-критичне та p-value:

In [14]:
t_stat = (x_bar - mu0) / (s / np.sqrt(n))
df = n - 1
t_crit = stats.t.ppf(1 - alpha, df)
p_value = 1 - stats.t.cdf(t_stat, df)

### Виводимо обчислення:

In [15]:
print (f'Середнє вибірки: {x_bar:.3f}')
print (f'Стандартне відхилення: {s:.3f}')
print (f't-статистика: {t_stat:.3f}')
print (f't-критичне: {t_crit:.3f}')
print (f'p-value: {p_value:.3f}')

Середнє вибірки: 132.286
Стандартне відхилення: 10.981
t-статистика: 2.069
t-критичне: 1.440
p-value: 0.042


### Висновок на основі t-значень:

In [18]:
if t_stat > t_crit:
  print('t-статистика більше за t-критичне. Відхиляємо гіпотезу H₀.')
else:
  print('t-статистика не більше за t-критичне. Не відхиляємо гіпотезу H₀.')

t-статистика більше за t-критичне. Відхиляємо гіпотезу H₀.


### Висновок на основі p-value:

In [19]:
if p_value < alpha:
  print('p-value менше за рівень значущості -> Відхиляємо гіпотезу H₀')
else:
  print('p-value не менше за рівень значущості -> Не відхиляємо гіпотезу H₀')

p-value менше за рівень значущості -> Відхиляємо гіпотезу H₀


In [23]:
print (f"""На рівні значущості {alpha} відхиляємо гіпотезу H₀.
Середній час до покупки на сайті збільшився після редизайну.""")

На рівні значущості 0.1 відхиляємо гіпотезу H₀. 
Середній час до покупки на сайті збільшився після редизайну.


### Чи зміниться ваше рішення при зміні рівня значущості на $\alpha = 0.05$?

Рішення не зміниться, ми так само відхиляємо гіпотезу H₀, але вже не так впевнено, так як p_value зовсім трохи менше за alpha -  0.042 < 0.05.

**Завдання 2.**

До спеціальної рекламної кампанії **23%** дорослих упізнавали логотип компанії. Після завершення кампанії відділ маркетингу провів опитування: з **1200** випадково відібраних дорослих **311** упізнали логотип.

Перевірте на рівні значущості **3%** ($\alpha=0.03$), чи дають ці дані достатні підстави стверджувати, що **тепер більше ніж 23%** дорослих упізнають логотип компанії. Для розвʼязку використовуйте бібліотеку `statsmodels`.

Зробіть висновок, чи зросла впізнаваність логотипу.

Додатково, обчисліть довірчий інтревал на заданому рівні значущості і проінтерпретуйте текстом - як він додатково пояснює прийняте нами рішення?

In [24]:
import statsmodels.api as sm
from statsmodels.stats.proportion import proportions_ztest, proportion_confint

In [25]:
# Дані:
count = 311
nobs = 1200
p0 = 0.23
alpha = 0.03

In [26]:
# z-тест для пропорції
stat, pval = proportions_ztest(count, nobs, p0, alternative='larger')
confint = proportion_confint(count, nobs, alpha, method='normal')

In [28]:
print('Результати z-тесту для пропорції:')
print(f'Z-статистика = {stat:.3f}')
print(f'p_value = {pval:.4f}')
print(f'97% довірчий інтервал: {confint}')

Результати z-тесту для пропорції:
Z-статистика = 2.306
p_value = 0.0106
97% довірчий інтервал: (0.23171700302179205, 0.28661633031154127)


### Інтерпретація результатів:

***1. Маємо правосторонній тест:***

 - H₀: p = 0.23

 - H₁: p > 0.23

 - Так як, p_value < alpha -> 0.0106 < 0.03 -> Відхиляємо H₀. Маємо достатньо підстав стверджувати, що більше ніж 23% дорослих впізнають логотип компанії.

***2. 97% довірчий інтервал (0.232, 0.287):***
 - З ймовірністю 97% справжня частка лежить у межах від 23.2% до 28.7%.


